xds/internal/xdsclient/xds.go

```diff
@@ -284,59 +277,61 @@ func processServerSideListener(lis *v3listenerpb.Listener) (*ListenerUpdate, err
+		case *v3listenerpb.Filter_TypedConfig:
+			// The typed_config field has an `anypb.Any` proto which could
+			// directly contain the serialized bytes of the actual filter
+			// configuration, or it could be encoded as a `TypedStruct`.
+			// TODO: Add support for `TypedStruct`.
+			tc := filter.GetTypedConfig()
+
+			// The only network filter that we currently support is the v3
+			// HttpConnectionManager. So, we can directly check the type_url
+			// and unmarshal the config.
+			// TODO: Implement a registry of supported network filters (like
+			// we have for HTTP filters), when we have to support network
+			// filters other than HttpConnectionManager.
+			if tc.GetTypeUrl() != version.V3HTTPConnManagerURL {
+				return nil, fmt.Errorf("filter chain {%+v} has unsupported network filter %q in filter {%+v}", filterChain, tc.GetTypeUrl(), filter)
 			}
-			if seenNames[name] {
-				return fmt.Errorf("filter chain {%+v} has duplicate filter name %q", filterChain, name)
+			hcm := &v3httppb.HttpConnectionManager{}
+			if err := ptypes.UnmarshalAny(tc, hcm); err != nil {
+				return nil, fmt.Errorf("filter chain {%+v} failed unmarshaling of network filter {%+v}: %v", filterChain, filter, err)
 			}
-			seenNames[name] = true
-
-			// Network filters have a oneof field named `config_type` where we
-			// only support `TypedConfig` variant.
-			switch typ := filter.GetConfigType().(type) {
-			case *v3listenerpb.Filter_TypedConfig:
-				// The typed_config field has an `anypb.Any` proto which could
-				// directly contain the serialized bytes of the actual filter
-				// configuration, or it could be encoded as a `TypedStruct`.
-				// TODO: Add support for `TypedStruct`.
-				tc := filter.GetTypedConfig()
-
-				// The only network filter that we currently support is the v3
-				// HttpConnectionManager. So, we can directly check the type_url
-				// and unmarshal the config.
-				// TODO: Implement a registry of supported network filters (like
-				// we have for HTTP filters), when we have to support network
-				// filters other than HttpConnectionManager.
-				if tc.GetTypeUrl() != version.V3HTTPConnManagerURL {
-					return fmt.Errorf("filter chain {%+v} has unsupported network filter %q in filter {%+v}", filterChain, tc.GetTypeUrl(), filter)
-				}
-				hcm := &v3httppb.HttpConnectionManager{}
-				if err := ptypes.UnmarshalAny(tc, hcm); err != nil {
-					return fmt.Errorf("filter chain {%+v} failed unmarshaling of network filter {%+v}: %v", filterChain, filter, err)
+			// "Any filters after HttpConnectionManager should be ignored during
+			// connection processing but still be considered for validity." - A36
+			if !seenHCM {
+				var err error
+				// "HTTPConnectionManager must have valid http_filters." - A36
+				if httpFilters, err = processHTTPFilters(hcm.GetHttpFilters(), true); err != nil {
+					return nil, fmt.Errorf("filter chain {%+v} had invalid server side HTTP Filters {%+v}", filterChain, hcm.GetHttpFilters())
```

Comment:
```go
filters, err := processHTTPFilters(...)
if err != nil {
	return nil, fmt.Errorf(...)
}
if !seenHCM {
	httpFilters = filters
	seenHCM = true
}
```

Reply:
Could you please address this comment as well. Thanks.

Reply:
Discussed in meeting. Decided to add validations but not persist the multiple HCM's.